Artemis integration (i.e. pretty blinking lights for your beep boop game)

skedgyedgy opened the pull request:
hi did you ever want your pretty keyboard lights to react to the stuff that happens in your funny beep boop man simulator rap battle rythm game well now you can!!!

-- added in an integration to artemis rgb which allows implementing rgb effects on basically any keyboard/mouse/whatever that allows that sort of control
-- stuff like the keyboard changing "backgrounds" to match with the level, changing the whole color during blammed, flashing during lightning strikes, fading to match the background colors of menus, etc. etc.
-- lua support
-- cool rainbow effect when you hold a full combo for long enough
-- fully customizable
-- o p t i o n a l
-- allows for custom trigger and values for modability and versatility and whatever

https://github.com/skedgyedgy/Artemis.Plugins.FNF/releases/tag/1.0
https://youtu.be/9A2IIk0EDN4

![Screenshot (42)](https://user-images.githubusercontent.com/96351114/147861233-ce13bf3c-688a-4845-bbc6-621482b3cbea.png)
![Screenshot (44)](https://user-images.githubusercontent.com/96351114/147861234-898ded51-a53e-41a8-9327-82a2e90a584c.png)
![Screenshot (45)](https://user-images.githubusercontent.com/96351114/147861235-2c4cb832-ebf7-4acd-b09e-8ced9b062b1b.png)
![Screenshot (46)](https://user-images.githubusercontent.com/96351114/147861236-36f0bb07-47c7-4a0b-b617-03770af27a6b.png)
![Screenshot (47)](https://user-images.githubusercontent.com/96351114/147861237-390fcb1b-a91c-47bf-9a64-5f41c32ba9cb.png)
![Screenshot (50)](https://user-images.githubusercontent.com/96351114/147861232-054ffc11-39b2-47d8-8af2-f5e3e4b5ef9d.png)

skedgyedgy committed:
yayy we can now connect to the

man this sure is a cool implementation it'd be a shame if it required jumping through forty thousand fucking hoops just to connect to the server once and confirm that it is in fact a server AHAHAHAHAHAHAHAHAHAH

skedgyedgy committed:
implemented sending player health to artemis integration

skedgyedgy committed:
you should stop spamming the logs... NOW!

skedgyedgy committed:
added background led colors to menus

skedgyedgy committed:
now sends healthbar colors to artemis

skedgyedgy committed:
absolutely necissary update

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
Merge branch 'ShadowMario:main' into main

skedgyedgy committed:
Merge branch 'ShadowMario:main' into artemis

skedgyedgy committed:
now sends a bunch more info to artemis module

skedgyedgy committed:
added blammed lights, accent colors and flashes to artemis integration

skedgyedgy committed:
added cool epic flashes

skedgyedgy committed:
game now sends title state on open and closed state on shutdown

skedgyedgy committed:
reset combo value on song start now

skedgyedgy committed:
added option to disable artemis in the visuals menu

skedgyedgy committed:
you should stop flooding the logs with a giant ass json string every time you connect to the artemis webserver.... NOW!

skedgyedgy committed:
menu state is now set when going into a menu,

skedgyedgy committed:
added fallback background colors in case there's nothing set in artemis profile

skedgyedgy committed:
added this cool epic magenta flash when you select something on the menu

skedgyedgy committed:
oops, that shouldn't have been there lol

skedgyedgy committed:
FUCK FUCK FUCK NO FUCK NO THAT DIDN'T WORK FUCK GO BACK UNDO UNDO

skedgyedgy committed:
added fade to black for transitions between menus/states

skedgyedgy committed:
Merge branch 'ShadowMario:main' into main

skedgyedgy committed:
Merge branch 'ShadowMario:main' into artemis

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
reworking how sending colors works 'cause what i'm trying to do right now is stupid

skedgyedgy committed:
woag lots of updates

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
added lua functions for artemis integration

also added custom event triggers for more versatility, and reworked how flashes and blammed lights are handled so they're now events instead of the weird way i did things before

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
oops that link should've been changed a while ago lol

skedgyedgy committed:
Merge pull request #4 from skedgyedgy/artemis

muchos colores

tposejank reviewed (approved):
sick af, don't think shadow mario will approve it though.

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy commented:
-- still a draft because i realized there's like a major improvement i could add to this that'd make it far easier to use for mods

ActualMandM commented on source/ArtemisIntegration.hx:
```diff
@@ -0,0 +1,255 @@
+using StringTools;
+
+class ArtemisIntegration {
+    // public static inline var ArtemisAPIUrlDirectory:String = "%ProgramData%/Artemis"; // WHAT DO YOU MEAN THIS DOESN'T WORK???????
```
You have to use `Sys.getEnv("ProgramData")` instead of `%ProgramData%`

skedgyedgy committed:
i may be stupid (thank you ActualMandM)

skedgyedgy committed:
HAHA IT WORKS (lots of words so go to desc)

implemented a new super cool system that lets mods just send profiles to artemis which means less user input required = thing is EVEN BETTER. this still needs to be cleaned up a lot plus there's still some functionality i wanna add but it's late so i'm gonna go to bed now

skedgyedgy commented:
-- i'm working on that thing i was talking about here: https://github.com/skedgyedgy/FNF-PsychEngine/tree/profile-control
-- basically allowing mods to include their own profile data and send them over themselves. i still need to flesh this out a lot more plus do some cleanup but it's late, you can take a look at what i've done so far if you'd like

skedgyedgy committed:
added (optional) id to mod metadata

skedgyedgy committed:
mods now send their id to artemis rather than their folder name (unless there is no id in which case folder name fallback lol)

skedgyedgy committed:
why is this imported twice go away

skedgyedgy committed:
make the number go up

skedgyedgy committed:
Merge pull request #5 from skedgyedgy/profile-control

Profile control

skedgyedgy commented:
now rather than the user having to install profiles, mods can include artemis profiles "MODFOLDER/artemis/default.json" which will automatically be added to artemis. there is also a vanilla profile ("assets/artemis/fnf-vanilla.json" by default) which can be edited by mod creators if they're actually editing the source code of the engine for their mod instead of a "drag-into-the-mods-folder" type of mod

skedgyedgy committed:
oops i accidentally broke vanilla let me fix that

skedgyedgy committed:
lua sethealth and addhealth functions now send health to artemis

bbpanzu commented:
cool idea. bookmarking

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into artemis

skedgyedgy committed:
implemented custom flags as well as new controley-displayey-on-keyboardey feature

(artemis api 1.2.x now)

skedgyedgy committed:
minor edit to the profile

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into artemis

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
made the number kinda sorta go up just for clarification

skedgyedgy committed:
set the number back for the pull request

skedgyedgy committed:
forgot to include "artemis" in the name of one of the older lua funcitons

skedgyedgy committed:
implemented a bunch of COOL EPIC AND AWESOME lua functions

skedgyedgy commented:
-- added a couple more of the things i wanted to add like the controls now changing on the keyboard to match your own custom controls instead of being stuck at WASD and arrow keys, and the mod can now control the colors of the control keys as well. this has also been done in a way which will support custom control such as 5 key or whatever since they work on a dictionary/map rather than just four variables named "left down up right"
-- also 8 custom booleans added to the integration that modders will be able to set that can be used for whatever they want

-- i'm getting close to completion. should i go through and surround all the points where artemis functions are called with "#if sys"? the functions are ignored if sys isn't supported since the "enabled" flag will never be set but idk if that'll still be necissary

skedgyedgy committed:
updated integration to support new features in api ver. 1.2.1

skedgyedgy committed:
painstakingly added "#if sys _ #end" around every single mere mention of artemis

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into artemis

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
Merge pull request #7 from skedgyedgy/artemis

more cool epic rgb stuff!!!!!!!!

skedgyedgy commented:
-- anyone wanna review this and make sure i didn't do anything majorly bad and awful? i'm still hesitant on taking this out of draft mode

skedgyedgy committed:
update vanilla artemis profile

set hints so it will now work with setups that aren't exactly my setup

skedgyedgy committed:
added a few little details

-- red danger border thing that was in the mallEvil stage now also shows on schoolEvil and when monster is being played
-- red flash whenever henchmen are killed :(
-- little effect plays when background ghouls spawn

skedgyedgy committed:
whoops

-- just realized the thing i did for monster would only trigger the first time you entered the song in story mode instead of, like, whenever you enter the song period

skedgyedgy committed:
Merge remote-tracking branch 'origin/artemis'

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into artemis

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
implemented api 1.3 stuff

skedgyedgy committed:
fixed a jarring transition effect

skedgyedgy committed:
oops

skedgyedgy committed:
oops again

Steviegt6 commented:
After some quick glances, your formatting is a little off (`functionName (parameters)` doesn't line up with the standard `functionName(parameters)` used in this repository, but that's pretty insignificant. The first major issue I see is a hardcoded port number - especially one that is by no means unique (9696). Would be safer to just scan for unoccupied ports and default to some really random port that no one is likely to already use. Other than that, it seems fine, but traces appear to be a little spammy and generally undescriptive.

skedgyedgy commented:
-- there is no hardcoded port number; there isn't even a server on fnf's side in the first place. that's just a default value i put in there during testing and left in, the first step of the program is to scan the "%ProgramData%/Artemis/webserver.txt" file for the actual address which replaces that value

skedgyedgy committed:
the red vignette effect i've been trying to implement to the monster stages for like a whole week now is finally working

skedgyedgy committed:
please spam the logs slightly less

skedgyedgy committed:
updated the url in options

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
added difficulty, song name, and character name changes to lua

skedgyedgy committed:
change character event now updates artemis

skedgyedgy committed:
added fnf icon to the vanilla profile

skedgyedgy committed:
extremely important addition

skedgyedgy committed:
Merge branch 'main' into fullscreen

skedgyedgy committed:
Merge pull request #11 from skedgyedgy/fullscreen

hihihihi can i alt+enter again pls thank u

skedgyedgy committed:
Merge pull request #12 from skedgyedgy/artemis

beep boop lights

skedgyedgy committed:
Revert "Merge pull request #12 from skedgyedgy/artemis"

This reverts commit 4e6f3efa90d1d43f408a7e2cd4a2259e1f0a5a28, reversing
changes made to 0053e6f897bdb7b16ea5b4b7ccff0816d87ec2a3.

skedgyedgy committed:
Revert "Merge pull request #11 from skedgyedgy/fullscreen"

This reverts commit 0053e6f897bdb7b16ea5b4b7ccff0816d87ec2a3, reversing
changes made to 4871a5a413f82c35a6ca1b5aff7bad163bc2b115.

XtraXD1 commented:
Good idea! I don't have one of these, but I like the idea. Hope that ShadowMario says the same :)
(Sorry for any grammar error lmao)

XtraXD1 commented:
Woo! So cool! What does it do? Recreate the screen, but in the keyboard? By the way, does it also color the custom keybinds? I mean, if you use "QWOP" to play, instead of "ASWD", if it colors QWOP of ASWD.

skedgyedgy commented:
@XtraXD1 -- thanks! also nope! i put in custom visuals for each of the vanilla stages, plus i'm making some for other mods too. i've done madness vandalization and am working on vs. sonic.exe next.

-- also yeah, the keybinds automatically update to match your own

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into artemis

skedgyedgy committed:
Merge branch 'main' into artemis

skedgyedgy committed:
ummmm... janky github moment???

github just decided that me merging my changes to here shouldn't actually merge them for whatever reason so uhhh lemme fix that

skedgyedgy committed:
added funny branch-specific text

MrClogsworthYT commented:
> @XtraXD1 -- thanks! also nope! i put in custom visuals for each of the vanilla stages, plus i'm making some for other mods too. i've done madness vandalization and am working on vs. sonic.exe next.
> 
> -- also yeah, the keybinds automatically update to match your own

why the hell do you you add -- to all your messages
like

-- hi

skedgyedgy commented:
-- 'cause it's comforting to me

HiroMizuki reviewed (approved):
oooo i forgor

HiroMizuki commented:
ik that maybe will not be merged, but i liked these

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main'

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into artemis

skedgyedgy committed:
Merge remote-tracking branch 'upstream/main' into skedgybranch

skedgyedgy committed:
Merge branch 'skedgybranch' into artemis

skedgyedgy committed:
oops let me fix that

skedgyedgy committed:
Merge branch 'skedgybranch' into artemis

SonicProYT commented:
make it compatible with 0.5.2h pls

Joalor64 reviewed (approved):
you need to be in the team this is amazing

ShadowMario commented:
That's cool and all but this integration could be an issue for mods and isn't really that relevant.
It sounds mostly like something that would exist in a mod instead of an engine itself.
